File: src/main/java/vocadb/notification/reader/configuration/SecurityConfiguration.java
package vocadb.notification.reader.configuration;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import java.nio.charset.StandardCharsets;
import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.io.buffer.DataBuffer;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.server.reactive.ServerHttpResponse;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.config.web.server.ServerHttpSecurity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.server.SecurityWebFilterChain;
import org.springframework.security.web.server.WebFilterExchange;
import org.springframework.security.web.server.authentication.AuthenticationWebFilter;
import org.springframework.security.web.server.util.matcher.ServerWebExchangeMatchers;
import org.springframework.web.server.ServerWebExchange;
import org.springframework.web.server.WebFilter;
import reactor.core.publisher.Mono;
import vocadb.notification.reader.security.CustomServerFormLoginAuthenticationConverter;
import static org.springframework.http.HttpStatus.FORBIDDEN;
import static org.springframework.http.HttpStatus.UNAUTHORIZED;

@Configuration(proxyBeanMethods = false)
@RequiredArgsConstructor
public class SecurityConfiguration {

    @Bean
    @SuppressWarnings("MultipleStringLiterals")
    public SecurityWebFilterChain securityWebFilterChain(final ServerHttpSecurity http) {
        SecurityWebFilterChain build = http.exceptionHandling(it -> {
            it.authenticationEntryPoint(this::authEntryPoint);
            it.accessDeniedHandler(this::accessDeniedHandler);
        })
            // TODO check/enable later
            .httpBasic().disable()
            .requestCache().disable()
            .csrf().disable()
            .cors().disable()
            .headers().disable()
            .anonymous().disable()

            .formLogin(e -> {
                e.loginPage("/api/login");
                e.requiresAuthenticationMatcher(
                    ServerWebExchangeMatchers.pathMatchers(HttpMethod.POST, "/api/login/authentication")
                );
                e.authenticationFailureHandler(this::failureHandler);
                e.authenticationSuccessHandler(this::successHandler);
            })
            .logout(e -> {
                e.logoutSuccessHandler(this::successHandler);
                e.logoutUrl("/api/logout");
            })

            .authorizeExchange(e -> {
                // permit all preflight requests
                e.pathMatchers(HttpMethod.OPTIONS).permitAll();

                // allow an arbitrary access to login and swagger endpoints
                e.pathMatchers("/api/login/**", "/api/logout").permitAll();

                // everything else must be authenticated
                e.anyExchange().authenticated();
            })
            .build();

        // currently formLogin builder does not allow to customize authentication converter
        build.getWebFilters().collectList().subscribe(
            webFilters -> {
                for (WebFilter filter : webFilters) {
                    if (filter instanceof AuthenticationWebFilter) {
                        AuthenticationWebFilter awf = (AuthenticationWebFilter) filter;
                        awf.setServerAuthenticationConverter(new CustomServerFormLoginAuthenticationConverter());
                    }
                }
            }
        );

        return build;
    }

    @SuppressFBWarnings("UP_UNUSED_PARAMETER")
    private Mono<Void> authEntryPoint(ServerWebExchange serverWebExchange, AuthenticationException exception) {
        return authErrorHandler(serverWebExchange, UNAUTHORIZED);
    }

    private String getErrorBody(HttpStatus unauthorized) {
        return String.format(
            "{\"status\": \"%d\", \"title\":\"%s\"}",
            unauthorized.value(), unauthorized.getReasonPhrase()
        );
    }

    @SuppressFBWarnings("UP_UNUSED_PARAMETER")
    private Mono<Void> accessDeniedHandler(ServerWebExchange serverWebExchange, AccessDeniedException e) {
        return authErrorHandler(serverWebExchange, FORBIDDEN);
    }

    @NotNull private Mono<Void> authErrorHandler(ServerWebExchange serverWebExchange, HttpStatus forbidden) {
        ServerHttpResponse response = serverWebExchange.getResponse();
        final String body = getErrorBody(forbidden);

        response.setStatusCode(forbidden);
        response.getHeaders().setContentType(MediaType.parseMediaType("application/problem+json"));

        DataBuffer buf = response.bufferFactory().wrap(body.getBytes(StandardCharsets.UTF_8));
        return response.writeWith(Mono.just(buf));
    }

    @SuppressFBWarnings("UP_UNUSED_PARAMETER")
    private Mono<Void> failureHandler(WebFilterExchange webFilterExchange, AuthenticationException exception) {
        return authErrorHandler(webFilterExchange.getExchange(), UNAUTHORIZED);
    }

    @SuppressFBWarnings("UP_UNUSED_PARAMETER")
    private Mono<Void> successHandler(WebFilterExchange webFilterExchange, Authentication authentication) {
        webFilterExchange.getExchange().getResponse().setStatusCode(HttpStatus.OK);
        return Mono.empty();
    }
}

File: src/main/java/vocadb/notification/reader/configuration/SecurityConfiguration.kt
package vocadb.notification.reader.configuration

import mu.KotlinLogging
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Import
import org.springframework.core.ResolvableType
import org.springframework.core.codec.Hints
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.codec.json.Jackson2JsonEncoder
import org.springframework.security.authentication.AccountStatusException
import org.springframework.security.authentication.AuthenticationServiceException
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurity
import org.springframework.security.config.web.server.ServerHttpSecurity
import org.springframework.security.core.Authentication
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.server.SecurityWebFilterChain
import org.springframework.security.web.server.WebFilterExchange
import org.springframework.security.web.server.authentication.RedirectServerAuthenticationEntryPoint
import org.springframework.security.web.server.authentication.ServerAuthenticationFailureHandler
import org.springframework.security.web.server.util.matcher.ServerWebExchangeMatchers
import org.springframework.stereotype.Component
import org.springframework.web.server.ServerWebExchange
import org.zalando.problem.Problem
import org.zalando.problem.Status
import org.zalando.problem.spring.webflux.advice.security.SecurityProblemSupport
import reactor.core.publisher.Mono
import vocadb.notification.reader.service.security.UserAuthority
import java.io.IOException

private val logger = KotlinLogging.logger { }

@EnableWebFluxSecurity
@Import(SecurityProblemSupport::class)
class SecurityConfiguration(
    val problemSupport: SecurityProblemSupport,
    val problemAuthenticationHandler: ServerAuthenticationFailureHandler
) {

    @Bean
    fun securityWebFilterChain(http: ServerHttpSecurity): SecurityWebFilterChain {
        return http
            .exceptionHandling {
                it.authenticationEntryPoint { exchange: ServerWebExchange, e: AuthenticationException ->
                    val redirect = RedirectServerAuthenticationEntryPoint("/login")

                    val request = exchange.request
                    val path = request.uri.path
                    when {
                        // redirect if request is made to non-api endpoints
                        !path.startsWith("/api") && !path.startsWith("/login/authentication") ->
                            redirect.commence(exchange, e)
                        else -> problemSupport.commence(exchange, e)
                    }
                }
                it.accessDeniedHandler(problemSupport)
            }
            .httpBasic().disable()
            // TODO check/enable later
            .requestCache().disable()
            .csrf().disable()
            .cors().disable()
            .headers().disable()

            .formLogin {
                it.loginPage("/login")
                it.requiresAuthenticationMatcher(
                    ServerWebExchangeMatchers.pathMatchers(
                        HttpMethod.POST,
                        "/login/authentication"
                    )
                )
                it.authenticationFailureHandler(problemAuthenticationHandler)
                it.authenticationSuccessHandler(this::successHandler)
            }
            .logout {
                it.logoutSuccessHandler(this::successHandler)
            }
            .authorizeExchange {
                it.pathMatchers(HttpMethod.OPTIONS, "/**").permitAll()
                it.pathMatchers("/js/**", "/css/**", "/img/**", "/favicon.ico").permitAll()
                it.pathMatchers("/login", "/login/**").permitAll()

                it.pathMatchers("/**").hasAuthority(UserAuthority.ROLE_USER)
                it.anyExchange().authenticated()
            }
            .build()
    }

    private fun successHandler(webFilterExchange: WebFilterExchange, authentication: Authentication?): Mono<Void> {
        webFilterExchange.exchange.response.statusCode = HttpStatus.OK
        return Mono.empty()
    }
}

@Component
class ProblemAuthenticationFailureHandler(
    val encoder: Jackson2JsonEncoder
) : ServerAuthenticationFailureHandler {

    override fun onAuthenticationFailure(
        webFilterExchange: WebFilterExchange,
        exception: AuthenticationException
    ): Mono<Void> {
        val response = webFilterExchange.exchange.response
        if (response.isCommitted) {
            logger.error {
                "Response is already committed. Unable to write exception " +
                        "${exception.javaClass.simpleName}: ${exception.localizedMessage}"
            }
            return Mono.empty()
        }

        val (status, problem) = exceptionMapper(exception)
        try {
            response.statusCode = status
            return toJson(problem, webFilterExchange.exchange)
        } catch (e: IOException) {
            logger.error(e) { "Failed to write error response" }
        }

        return Mono.empty()
    }

    private fun toJson(body: Any, exchange: ServerWebExchange) =
        exchange.response.writeWith(
            encoder.encode(
                Mono.just(body),
                exchange.response.bufferFactory(),
                ResolvableType.forInstance(body),
                MediaType.APPLICATION_JSON,
                Hints.from(Hints.LOG_PREFIX_HINT, exchange.logPrefix)
            )
        )

    private fun exceptionMapper(exception: AuthenticationException): Pair<HttpStatus, Problem> {
        val problem: Problem = when (exception) {
            is AuthenticationServiceException -> Problem.valueOf(Status.INTERNAL_SERVER_ERROR)
            is AccountStatusException -> Problem.valueOf(
                Status.FORBIDDEN,
                exception.message ?: "User is either disabled or has expired credentials"
            )
            is BadCredentialsException -> Problem.valueOf(
                Status.UNAUTHORIZED,
                "Invalid credentials"
            )
            else -> Problem.valueOf(Status.UNAUTHORIZED)
        }
        val status = when (exception) {
            is AuthenticationServiceException -> Status.INTERNAL_SERVER_ERROR
            is AccountStatusException -> Status.FORBIDDEN
            else -> Status.UNAUTHORIZED
        }

        return Pair(HttpStatus.valueOf(status.statusCode), problem)
    }
}
